lib/cudadrv/stream.jl

```diff
@@ -153,13 +153,31 @@ end
         spins += 1
     end
 
-    # minimize CPU usage of long-running kernels
-    # by waiting for an event signalled by CUDA
-    event = Threads.Event()
+    # minimize CPU usage of long-running kernels by waiting for an event signalled by CUDA
+    event = Base.Event()
     launch(; stream) do
         notify(event)
     end
-    Base.wait(event)
+    # if an error occurs, the callback may never fire, so use a timer to detect such cases
+    dev = device()
+    timer = Timer(0; interval=1)
+    Base.@sync begin
+        Threads.@spawn begin
+            device!(dev)
+            while isopen(timer)
+                if Base._trywait(timer) # to prevent EOFError
+                    if unsafe_cuStreamQuery(stream) != ERROR_NOT_READY
+                        notify(event)
+                    end
+                end
+            end
+        end
```

Comment:
Maybe you have some reasons to use the internal function, but, this looks like implementable only with public API:

```suggestion
        Threads.@spawn begin
            device!(dev)
            while true
                try
                    wait(timer)
                catch err
                    err isa EOFError && break
                    rethrow()
                end
                if unsafe_cuStreamQuery(stream) != ERROR_NOT_READY
                    notify(event)
                    break  # why not?
                end
            end
        end
```

I'd also probably call `notify(event)` in a `finally` close so that the tasks "unstuck" when some unexpected errors happen.

```suggestion
        Threads.@spawn try
            device!(dev)
            while true
                try
                    wait(timer)
                catch err
                    err isa EOFError && break
                    rethrow()
                end
                if unsafe_cuStreamQuery(stream) != ERROR_NOT_READY
                    break
                end
            end
        finally
            notify(event)
        end
```

Reply:
I wanted to avoid the exception handling, for performance. But maybe that's overly cautious / not worth using an internal API for?
The alternative is `Timer() do ... done`, but then we'd do a `device!` call at every iteration, which is bad for performance.

Reply:
Yeah, that's exactly why I wanted https://github.com/JuliaLang/julia/discussions/43773  for things like a public `trywait` function and more. I thought the performance problem would be minimal since you are using the timer with a 1-second interval. But it's kind of a "meh" point to suggest to stick with public API in CUDA.jl (that is copuled with `julia` compiler internals) anyway so feel free to ignore my comment :)

Reply:
Yeah you're right, I went with the exception-based approach. Thanks for the suggestion!